### PR TITLE
Creating image links #3186

### PR DESCRIPTION
**Pull request by Mr-Magnificent**
Signed-off-by: Ayush Poddar <ayush_poddar@live.com>

#### What does this PR do?
Images within Anchor acquire the size of the container
#### What testing has been done on this PR?
Storybook
#### What are the relevant issues?
#3186 

#### Screenshots (if appropriate)
Screenshot_20191008_202840

#### Do the grommet docs need to be updated?
Maybe
#### Should this PR be mentioned in the release notes?
NA
#### Is this change backwards compatible or is it a breaking change?
NA

### CODE REVIEW

**Comment by ShimiSun** [comment]
Thank you for filing this PR and addressing this issue!
I'm thinking whether those styling should be applied directly as you suggested or via an additional `fill` prop.
Adding those styling to existing used styles might impact current users and I'm not sure if it will be backward compatible in those cases.
I'd love a second opinion on this one @ericsoderberghp.

**Comment by oorestisime** [comment]
Yeap was thinking the same with @ShimiSun here. Probably will make images stretched if the container is bigger than the image and this won't be desired by users (or at least would be a breaking change) so i am inclining for a fill prop as well.
Nice investigation and thanks for working on this

**Comment by Mr-Magnificent** [comment]
If possible, I'd love to continue working on this but I'd need some help and directions on how to implement the above method. Maybe point to an example in the codebase? :smile:

**Comment by oorestisime** [comment]
For sure! We are not taking over :)

So you can have a look at Paragraph component it has recently acquired the fill prop :)
I think fill should be by default false for backwards compatibility.

Generally the files to change should be:

- doc.js for the documentation and prop types
- index.d.ts for the typescript
- Image file to pass in the fill prop to styled image
- styled image file to take into account the fill prop and do something along the following https://github.com/grommet/grommet/blob/master/src/js/components/Paragraph/StyledParagraph.js#L16

We would also love if you can can add a test in the tests folder making use of the fill prop. 

If you have any question feel free to ask here or better join the slack channel :)

**Comment by ShimiSun** [comment]
Thank you for owning this one @Mr-Magnificent, as @oorestisime mentioned above, this PR will be a useful reference for you https://github.com/grommet/grommet/pull/3298
Let us know if you need any further help here!

**Comment by Mr-Magnificent** [comment]
Actually, I was preoccupied with some college work. Will be a couple of days before I can jump back in. I hope that is not a problem.

**Comment by ShimiSun** [comment]
@Mr-Magnificent sure thing, let us know if you need any further guidance when you are back at it.

**Comment by ericsoderberghp** [comment]
I think I am missing something, why isn't the `fit` property sufficient to address this?

**Comment by ShimiSun** [comment]
@ericsoderberghp Interesting comment.
I took a look at `object-fit` CSS, and we might want to enhance our `fit` prop. 
At the moment `fit` supports `cover` and `contain` but it doesn't support `fill`, which might answer this issue. 
The difference between `cover` and `fill` of the `fit` values is
>cover
>The replaced content is sized to maintain its aspect ratio while filling the element’s entire content box. If the object's aspect ratio does not match the aspect ratio of its box, then the object will be clipped to fit.
>fill
>The replaced content is sized to fill the element’s content box. The entire object will completely fill the box. If the object's aspect ratio does not match the aspect ratio of its box, then the object will be stretched to fit.

https://developer.mozilla.org/en-US/docs/Web/CSS/object-fit

**Comment by ericsoderberghp** [comment]
I'm just wondering when `fill` might be used, since it doesn't preserve the aspect ratio.

**Comment by Mr-Magnificent** [comment]
From what I could test out from the codebase `object-fit` doesn't seem to work because the `<a>` which wraps the image because of `<Anchor>` in the DOM acquires the complete size of the image because of it's inline-nature and all of the `object-fit` seems to turn non-functional.

Screenshot_20191015_132625

File: src/js/components/Image/doc.js [hunk 1]
```diff
@@ -26,6 +26,9 @@ export const doc = Image => {
       PropTypes.string,
       PropTypes.bool,
     ]).description('Transparency of the image.'),
+    fill: PropTypes.bool.description(
+      'Set maximum width to that of parent element',
+    ),
```
Comment (ShimiSun) [inline src/js/components/Image/doc.js]: I think you should mention both width and height.

File: src/js/components/Image/doc.js [hunk 2]
```diff
@@ -26,6 +26,9 @@ export const doc = Image => {
       PropTypes.string,
       PropTypes.bool,
     ]).description('Transparency of the image.'),
+    fill: PropTypes.bool.description(
```
Comment (ShimiSun) [inline src/js/components/Image/doc.js]: I'm inclining for consistency of the fill prop.
When `fill` is defined on other components it allows the values of 'vertical' | 'horizontal' | boolean.
It's been done a few days ago in a similar way for WorldMap.
Do you feel like it is needed for Image as well?

**Comment by ShimiSun** [comment]
@Mr-Magnificent thank you for following up on this!
I'm not sure I understand the image you uploaded, is this image created from your branch and uses the `fill` for image?
Regardless, I think it is worth asking the question of whether we want to have the `fill` as an independent prop as you suggested, or is it more semantically accurate to add it as on option for `fit` since internally being used as part of `object-fit`.

**Comment by Mr-Magnificent** [comment]
Yes, I had made the changes to see if `object-fit = fill` worked in my local branch by adding the necessary code. It can be seen that `object-fit` was indeed applied on the image by checking the `CSS pane`, but the image still overflowed the container (highlighted).

Yes, I do agree about `fill` and `fit` since both of them seem so similar and I wasn't even sure about the wordings to put in the docs since they try to achieve similar things logically. I would be glad if you could point out the course of action.

File: src/js/components/Image/doc.js (hunk 2 above)
Comment (Mr-Magnificent) [reply src/js/components/Image/doc.js]: Sure thing! I had referred to [https://github.com/grommet/grommet/issues/3222#issuecomment-516176939](url) to assume that `fill` should assumes only `true` or `false`. Should all the values be supported?

**Comment by ShimiSun** [comment]
I think that as much as it will be consistent with other components to use `fill` as a prop, it will be more semantically accurate to use it as one of `fit` options. but I'd love a second opinion cause I'm not that convinced either...

File: src/js/components/Image/StyledImage.js [hunk 3]
```diff
@@ -14,9 +14,15 @@ const fitStyle = css`
   object-fit: ${props => FIT_MAP[props.fit]};
 `;
 
+const fillStyle = css`
+  max-width: 100%;
```
Comment (ericsoderberghp) [inline src/js/components/Image/StyledImage.js]:
```suggestion
  width: 100%;
```

File: src/js/components/Image/StyledImage.js [hunk 4]
```diff
@@ -14,9 +14,15 @@ const fitStyle = css`
...
+  max-height: 100%;
```
Comment (ericsoderberghp) [inline src/js/components/Image/StyledImage.js]:
```suggestion
  height: 100%;
```

File: src/js/components/Image/__tests__/__snapshots__/Image-test.js.snap [hunk 5]
```diff
@@ -1,5 +1,32 @@
...
+  max-width: 100%;
+  max-height: 100%;
+}
+
+<div
+  className="c0"
+>
+  <img
+    className="c1"
+    fill={true}
```
Comment (ericsoderberghp) [inline src/js/components/Image/__tests__/__snapshots__/Image-test.js.snap]: We shouldn't be leaking the `fill` into the DOM. We handle this in other areas by renaming the prop in Image to something like `fillProp` and passing that to StyledImage.

File: src/js/components/Image/doc.js (hunk 2 above)
Comment (ericsoderberghp) [reply src/js/components/Image/doc.js]: I think we're OK to just have it be a boolean for now.

File: src/js/components/Image/doc.js [hunk 6]
```diff
@@ -26,6 +26,9 @@ export const doc = Image => {
...
+      'Set maximum width to that of parent element',
```
Comment (ericsoderberghp) [inline src/js/components/Image/doc.js]:
```suggestion
      'Whether the image expands to fill all of the available width and height.',
```

File: src/js/components/Image/stories/Fill.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    <Grommet theme={grommet}>
+      <Box align="start" gap="small">
+        <Box height="small" width="small" border>
+          <Anchor href="#">
+            <Image fit="cover" fill src="//v2.grommet.io/assets/IMG_4245.jpg" />
+          </Anchor>
+        </Box>
+        <Box height="small" width="small" border>
+          <Anchor href="#">
+            <Image fit="cover" src="//v2.grommet.io/assets/IMG_4245.jpg" />
```
Comment (ericsoderberghp) [inline src/js/components/Image/stories/Fill.js]:
```suggestion
            <Image fit="contain" fill src="//v2.grommet.io/assets/IMG_4245.jpg" />
```

**Comment by ShimiSun** [comment]
hi @Mr-Magnificent, will you be able to follow up on the last small comments? we appreciate your work on this issue.

**Comment by Mr-Magnificent** [comment]
Yeah, Sorry was just slightly preoccupied with some work. I have made the reviewed changes. Kindly check.

File: src/js/components/Image/__tests__/__snapshots__/Image-test.js.snap (hunk 5 above)
Comment (ShimiSun) [reply src/js/components/Image/__tests__/__snapshots__/Image-test.js.snap]: @Mr-Magnificent we shouldn't have our props to appear on the DOM in order to keep the DOM nice and clean, can you please take a look at this last comment?

**Comment by Mr-Magnificent** [comment]
Hi @ShimiSun, I was wondering whether these changes are okay?

**Comment by ShimiSun** [comment]
Yep, @Mr-Magnificent , thank you for your friendly reminder. reviewing again now

File: src/js/components/Image/doc.js [hunk 8]
```diff
@@ -26,6 +26,9 @@ export const doc = Image => {
       PropTypes.string,
       PropTypes.bool,
     ]).description('Transparency of the image.'),
+    fillProp: PropTypes.bool.description(
```
Comment (ShimiSun) [inline src/js/components/Image/doc.js]: I think the prop should still be called `fill` for the users (consistent across grommet) and it should use the `fillProp` internally as an alias.

**Comment by Mr-Magnificent** [comment]
@ShimiSun kindly check whether this is okay?

**Comment by Mr-Magnificent** [comment]
Hi @ShimiSun, Does this require anything else too?

**Comment by ShimiSun** [comment]
Thank you for your contribution! 
This change is included on grommet 2.8.1 🎉

**Pull request merged**